20240328-queries-with-usersets.md

```diff
@@ -0,0 +1,176 @@
+| -                                   | Check(document:1#union@document:1#a)                      | `{allowed=false}` | `{allowed=true}`              |
+| -                                   | Check(document:1#union@document:1#b)                      | `{allowed=false}` | `{allowed=true}`              |
+| document:1#parent:group:marketing   | Check(document:1#tuple_to_userset@group:marketing#member) | `{allowed=false}` | `{allowed=true}`              |
+| -                                   | Check(document:1#intersection@document:1#a)               | `{allowed=false}` | `{allowed=false}` (no change) |
+| -                                   | Check(document:1#intersection@document:1#b)               | `{allowed=false}` | `{allowed=false}` (no change) |
+| -                                   | Check(document:1#difference_1@document:1#a)               | `{allowed=false}` | `{allowed=false}` (no change) |
+| -                                   | Check(document:1#difference_1@document:1#a)               | `{allowed=false}` | `{allowed=false}` (no change) |
+| document:1#c@group:marketing#member | Check(document:1#difference_2@group:marketing#member)     | `{allowed=true}`  | `{allowed=true}`  (no change) |
+
+
+https://play.fga.dev/stores/create/?id=01HT3QDCENG2D0A3KS0CNQSTAJ.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Check API involves solving a parent problem that may contain subproblems. The implementation of this proposal will be that when solving the parent problem or each subproblem, if the problem is of the form `objectType:objectId # relation @ objectType:objectId#relation`, then we will immediately return `{allowed=true}`.
+
+The same idea will apply to ListObjects and ListUsers APIs.
+
+# Migration
+[migration]: #migration
+
+This proposal doesn't entail any migration of data. However, it will require a thorough review by application developers and authorization model authors to ensure that their applications are making the right queries to OpenFGA. Some developers making Check API calls with a userset in the `user` field may want to change this to be a specific user, as the application may start granting more permissions than desired.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Confusion for application developers. So far, developers have largely assumed that a Check returns `{allowed: true}` if and only if there is a tuple or tuples in the system saying so. With this proposal we are introducing the idea of "implicit" tuples. 
+
+# Alternatives
+[alternatives]: #alternatives
+
+N/A
+
+# Prior Art
+[prior-art]: #prior-art
+
+N/A
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Given that the userset `document:1#viewer` is defined as the set that has the `viewer` relation with `document:1`, should we block application developers from writing the tuple `document:1#viewer@document:1#viewer` since the Write will become irrelevant? And if yes, what error should we return?
```

Comment:
Can we include this in scope? If we are introducing implicit tuples, still allowing them to be written will be both superfluous and confusing (it allows devs to assume they can delete these tuples).

A new error probably, but it should mimic the fact that you cannot write duplicates.

Maybe `writing tuples that are implicit is not allowed` or something along those lines?

Reply:
I think this train of thought is misleading. We're not introducing any notion of tuples or changes to tuple semantics. This reflexive property is independent of that.

The statements/behavior around type restrictions and how that pertains to what can be written in the Write API remains the same. Nothing should change with Read or Write as a side-effect of this change in FGA query behavior. It's query behavior, not write or read behavior.